### PR TITLE
Auto start stream on viewer and auto stop stream on no viewer.

### PR DESCRIPTION
**Pull request by lastpeony**
If configured allows streams to auto start when there is a viewer trying to watch offline stream. 
If configured allows streams to auto stop after configured time elapses when there is no viewer.

Auto start will only work for IP Camera, Stream Source, VOD and playlist stream types on HLS,WebRTC and DASH.

Todo:
work on integration tests and cluster support -- done

https://github.com/ant-media/Ant-Media-Server/issues/4404

### CODE REVIEW

**Comment by asharamseervi** [comment]
Hi @lastpeony, This is an absolutely great add-on to AMS! Though need to clarify, if we had taken care of a few concerns as mentioned below:

Can you please help me to understand, how the stream session will behave when there is no viewer?

Let's say any user enabled recording streams from configurations and started streaming their session. While few audiences joined and this live-streaming session is meant to be of 1 hour. And, let's say there is no user after 30 minutes:
- Does steam will continue to work till the session end?
- Will recording will continue to record till the session completes?
- Will adaptive streaming will also manage to run till the session completes?

Vice-versa, if no audience joined the session while initiating the stream session. The session will continue to run for 1 hour flawlessly?

**Comment by lastpeony** [comment]
Hi @asharamseervi 
Its possible to configure time elapse to stop after no viewer detected.
we get the time when there is no viewer.
User set 30 seconds
Stream stops after 30 seconds when there is no viewer detected.

so in your example if user sets 0 seconds stream will stop.
but if user sets 30 minutes stream will continue to play for 30 more minutes after no viewer detected.

this feature currently stops the stream so i dont think recording will continue.
yes if stream is not stoped it no changes related to adaptive bitrate streaming.

**Comment by asharamseervi** [comment]
Thanks @lastpeony for updating! Looks good to me.

File: src/main/java/io/antmedia/filter/WaitingViewerFilter.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+public class WaitingViewerFilter extends AbstractFilter{
+
+
+    protected static Logger logger = LoggerFactory.getLogger(WaitingViewerFilter.class);
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        final HttpServletRequest httpRequest =(HttpServletRequest)request;
+        final String method = httpRequest.getMethod();
+        if(HttpMethod.HEAD.equals(method) && (httpRequest.getRequestURI().endsWith("m3u8") || httpRequest.getRequestURI().endsWith("m4s"))){
```
[the remark this answers is not in the document]
Comment (lastpeony) [reply src/main/java/io/antmedia/filter/WaitingViewerFilter.java:26]: @sonatype-lift ignoreall

**Comment by timantmedia** [comment]
Customer wants this feature FD108089 @lastpeony @mekya @burak-58 how likely is this going to be merged?

**Pull request merged**